// @zxcvbn-ts/core/dist/Matching.js
'use strict';

var helper = require('./helper.js');
var matching = require('./matcher/date/matching.js');
var matching$1 = require('./matcher/dictionary/matching.js');
var matching$2 = require('./matcher/regex/matching.js');
var matching$3 = require('./matcher/repeat/matching.js');
var matching$4 = require('./matcher/sequence/matching.js');
var matching$5 = require('./matcher/spatial/matching.js');
var Options = require('./Options.js');

class Matching {
  constructor() {
    this.matchers = {
      date: matching,
      dictionary: matching$1,
      regex: matching$2,
      repeat: matching$3,
      sequence: matching$4,
      spatial: matching$5
    };
  }

  match(password) {
    const matches = [];
    const matchers = [...Object.keys(this.matchers), ...Object.keys(Options.matchers)];
    matchers.forEach(key => {
      if (!this.matchers[key] && !Options.matchers[key]) {
        return;
      }

      const Matcher = this.matchers[key] ? this.matchers[key] : Options.matchers[key].Matching;
      const usedMatcher = new Matcher();
      helper.extend(matches, usedMatcher.match({
        password,
        omniMatch: this
      }));
    });
    return helper.sorted(matches);
  }

}

module.exports = Matching;
